### PR TITLE
Combat AI: enable defensive tactics

### PR DESCRIPTION
**Pull request by idshibanov**
Adding protective mode for the units (ranged unit protection) including castle defense.
Small fixes here and there.

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 1]
```diff
@@ -285,26 +285,70 @@ namespace AI
             }
         }
         else {
+            // Melee unit decision tree (both flyers and walkers)
+
             if ( defensiveTactics ) {
                 // Melee unit - Defensive action
+                const double defendDistanceModifier = myArmyStrength / STRENGTH_DISTANCE_FACTOR;
+                double maxArcherValue = defendDistanceModifier * ARENASIZE * -1;
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:293]: Please modify into `double maxArcherValue = -defendDistanceModifier * ARENASIZE;`

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 2]
```diff
@@ -285,26 +285,70 @@ namespace AI
...
+                double maxEnemyThreat = 0;
+
+                for ( const Unit * unitToDefend : friendly ) {
+                    if ( unitToDefend->GetUID() != currentUnit.GetUID() && unitToDefend->isArchers() ) {
+                        const int headIndexToDefend = unitToDefend->GetHeadIndex();
+                        std::pair<int, uint32_t> move = arena.CalculateMoveToUnit( *unitToDefend );
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:299]: Please make it `const`.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 3]
```diff
@@ -285,26 +285,70 @@ namespace AI
...
+
+                        DEBUG( DBG_AI, DBG_TRACE, unitToDefend->GetName() << " archer value " << archerValue << " distance: " << distanceToUnit );
+
+                        // Search for enemy units threatening our archers within range
+                        const Indexes & adjacentEnemies = Board::GetAdjacentEnemies( *unitToDefend );
+                        for ( const int cell : adjacentEnemies ) {
+                            const Unit * enemy = Board::GetCell( cell )->GetUnit();
+                            if ( enemy ) {
+                                const double enemyThreat = enemy->GetScoreQuality( currentUnit );
+                                std::pair<int, uint32_t> moveToEnemy = arena.CalculateMoveToUnit( *enemy );
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:311]: Please make it `const`.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 4]
```diff
@@ -285,26 +285,70 @@ namespace AI
...
+                                const bool canReach = moveToEnemy.first != -1 && moveToEnemy.second <= currentUnitMoveRange;
+                                const bool hadAnotherTarget = target != NULL;
+
+                                // Composite priority criteria:
+                                // 1. Enemy is within move range
+                                // 2. Archer unit value
+                                // 3. Enemy unit threat
+                                if ( canReach != hadAnotherTarget && canReach
+                                     || ( canReach == hadAnotherTarget
+                                          && ( maxArcherValue < archerValue || ( maxArcherValue == archerValue && maxEnemyThreat < enemyThreat ) ) ) ) {
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:321]: Please add parentheses () as it's not obvious where logical OR belong to: ` if ( a && b || c )`

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 5]
```diff
@@ -285,26 +285,70 @@ namespace AI
...
+                else if ( targetCell != -1 ) {
+                    DEBUG( DBG_AI, DBG_INFO, currentUnit.GetName() << " protecting friendly archer, moving to " << targetCell );
                 }
             }
             else {
                 // Melee unit - Offensive action
-                const double distanceModifier = enemyArmyStrength / STRENGTH_DISTANCE_FACTOR;
-                double maxPriority = distanceModifier * ARENASIZE * -1;
+                const double attackDistanceModifier = enemyArmyStrength / STRENGTH_DISTANCE_FACTOR;
+                double maxPriority = attackDistanceModifier * ARENASIZE * -1;
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:351]: Please modify it into `double maxPriority = -attackDistanceModifier * ARENASIZE;`

**Review by ihhub — COMMENTED** [review]
Hi @idshibanov , this pull request looks good. I've left just few minor comments. Please take a look into them.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 4 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:321]: Where? There's already 4 parentheses...

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:321]: `if ( ( canReach != hadAnotherTarget && canReach ) || ( canReach == hadAnotherTarget ... ) )`. Before first ||

**Comment by ihhub** [comment]
Hi @idshibanov , good job! Thanks a lot for AI enhancements!

**Pull request merged**